cmd/tether/attach.go

```diff
@@ -150,68 +178,102 @@ func (t *attachServerSSH) start() error {
 func (t *attachServerSSH) stop() {
 	defer trace.End(trace.Begin("stop attach server"))
 
-	if t == nil || !t.enabled {
+	if t == nil || !t.IsEnabled() {
 		return
 	}
 
-	t.enabled = false
-	conn := t.conn
-	t.conn = nil
+	log.Debugf("Setting enabled to false")
+	t.Disabled()
 
-	if conn != nil {
-		(*conn).Close()
+	log.Debugf("Canceling the context")
+	t.cancel()
+
+	log.Debugf("Acquiring the connection lock")
+	t.conn.Lock()
+	if t.conn.conn != nil {
+		log.Debugf("Closing the connection")
+		(*t.conn.conn).Close()
+		t.conn.conn = nil
 	}
+	t.conn.Unlock()
+	log.Debugf("Released the connection lock")
 }
 
 // run should not be called directly, but via start
 // run will establish an ssh server listening on the backchannel
 func (t *attachServerSSH) run() error {
 	defer trace.End(trace.Begin("main attach server loop"))
 
-	var serverConn *ssh.ServerConn
+	var serverConn struct {
+		sync.Mutex
+		*ssh.ServerConn
+	}
+	var established bool
+
 	var chans <-chan ssh.NewChannel
 	var reqs <-chan *ssh.Request
 	var err error
 
-	for t.enabled {
+	for t.IsEnabled() {
+		serverConn.Lock()
+		established = serverConn.ServerConn != nil
+		serverConn.Unlock()
+
 		// keep waiting for the connection to establish
-		for serverConn == nil {
+		for !established && t.IsEnabled() {
+			log.Debugf("Trying to establish a connection")
 			// tests are passing their own connections so do not create connections when testing is set
-			if !t.testing {
-				// close the connection if required
-				if t.conn != nil {
-					(*t.conn).Close()
+
+			establishFn := func() error {
+				t.conn.Lock()
+				defer t.conn.Unlock()
```

Comment:
`for` loop with a `defer Unlock`.  You'll pile these `defer`s up if you don't exit the loop, right?